[ADDED] FileStore: option to limit use of File Descriptors

kozlovic opened the pull request:
Each channel corresponds to a sub-directory and each
sub-directory contain several files. The more channels are
added to the system, the more FDs are needed.

This option is a soft limit instructing the store to
try to keep the number of opened files below this limit.
It is possible that during concurrent read/write across
multiple channels the number of FDs goes above the limit.

Note that this is for the file store's FDs, not the server
as a whole. For instance, if the server embeds the NATS
server, the number of FDs for the nats-streaming-server
process could be much higher (additional FD per Streaming
client connecting to the server). If that's an issue, one
should configure the nats-streaming-server so that it
connects to a remote NATS Server.

kozlovic commented on stores/filestore.go:
```diff
@@ -289,12 +305,49 @@ const (
+	rootDir string
+	files   map[fileID]*file
+	nextID  fileID
+	closed  bool
```
done

kozlovic commented on stores/filestore.go:
```diff
@@ -405,20 +456,20 @@ type FileMsgStore struct {
-	currSlice    *fileSlice
-	rootDir      string
+	writeSlice   *fileSlice
+	channel      string
```
done

kozlovic commented on stores/filestore.go:
```diff
@@ -665,6 +715,286 @@ func (w *bufferedWriter) checkShrinkRequest() {
+// the requested file. If the system's file descriptor limit is reached,
+// opening the file will fail and that error will be returned to the caller.
+// Lock is required on entry.
+func (fm *filesManager) _closeUnusedFiles(idToSkip fileID) {
```
Yes, you got the meaning. But it is not a new convention that I would impose on the group. I would also welcome a better way to do this (without making `filesManager` public).

ColinSullivan1 replied:
No such option I'm aware of without refactoring into a new module.  IMO they are fine without the `_`, but just an opinion.

kozlovic commented on stores/filestore.go:
```diff
@@ -665,6 +715,286 @@ func (w *bufferedWriter) checkShrinkRequest() {
+}
+
+// remove a file from the list of files. The initial state must be either `fileOpened`
+// or `fileClosed`. This call will loop until it can switch the file's state from
```
This function is calling the function above. So yes, it will loop for a certain number of attempts if the file is not in the state it should be.

kozlovic commented on stores/filestore.go:
```diff
@@ -2041,19 +2468,16 @@ func (ms *FileMsgStore) backgroundTasks() {
-						opened--
-					}
+				// If no FD limit and this is the write slice, skip.
+				if !ms.hasFDsLimit && slice == ms.writeSlice {
```
The line you refer to optimizes the fact that there is no need to notify (through the setting of the atomic variable) the background task that slices should be checked if the given slice is the write slice. The code here is simply saying that when check is happening, we should not close the writeSlice if there is no FD limit.

kozlovic commented on stores/filestore.go:
```diff
@@ -441,25 +492,24 @@ var (
 		checkVersion = true
 	}
-	file, err := os.OpenFile(fileName, mode, 0666)
+	file, err := os.OpenFile(fileName, flags, 0666)
```
You mean 0600? You would want the user to be able to read/write the file.. I would suggest that we take that to a different PR and likely have a `FileStoreOption` that allows one to specify the flag they want?

ColinSullivan1 replied:
Yep, flipped them around.  User = r/w, group = r, other = nothing.  A different PR, allowing the user to supply a mask sounds good here.

kozlovic commented:
@ColinSullivan1 I have removed the `_` prefix from the "private" functions and squashed the other update based on comment to a single commit for the merge. Note that comments to squashed commits are still avail in GitHub (clicking in `Show outdated` in places where there was a comment that is no longer relevant due to the squash). Thanks for the review!
